things/cache/things_test.go

```
@@ -0,0 +1,201 @@
+			key:  testKey2,
+			id:   testID2,
+			err:  nil,
+		},
+		{
+			desc: "Save thing with long key ",
+			key:  strings.Repeat("a", 513*1024*1024),
+			id:   testID,
+			err:  errors.ErrCreateEntity,
+		},
+		{
+			desc: "Save thing with long id ",
+			key:  testKey,
+			id:   strings.Repeat("a", 513*1024*1024),
+			err:  errors.ErrCreateEntity,
+		},
+		{
+			desc: "Save thing with empty key",
+			key:  "",
+			id:   testID,
+			err:  nil,
+		},
+		{
+			desc: "Save thing with empty id",
+			key:  testKey,
+			id:   "",
+			err:  nil,
+		},
+		{
+			desc: "Save thing with empty key and id",
+			key:  "",
+			id:   "",
+			err:  nil,
+		},
+	}
+
+	for _, tc := range cases {
+		err := tscache.Save(ctx, tc.key, tc.id)
+		if err == nil {
+			id, _ := tscache.ID(ctx, tc.key)
+			assert.Equal(t, tc.id, id, fmt.Sprintf("%s: expected %s got %s", tc.desc, tc.id, id))
+		}
+		assert.True(t, errors.Contains(err, tc.err), fmt.Sprintf("%s: expected %s got %s", tc.desc, tc.err, err))
+	}
+}
+
+func TestID(t *testing.T) {
+	redisClient.FlushAll(context.Background())
+	tscache := cache.NewCache(redisClient, 1*time.Minute)
+	ctx := context.Background()
+
+	err := tscache.Save(ctx, testKey, testID)
+	assert.Nil(t, err, fmt.Sprintf("Unexpected error while trying to save: %s", err))
+
+	err = tscache.Save(ctx, testKey2, "")
+	assert.Nil(t, err, fmt.Sprintf("Unexpected error while trying to save: %s", err))
+
+	err = tscache.Save(ctx, "", testID2)
+	assert.Nil(t, err, fmt.Sprintf("Unexpected error while trying to save: %s", err))
+
+	cases := []struct {
+		desc string
+		key  string
+		id   string
+		err  error
+	}{
+		{
+			desc: "Get thing ID from cache",
+			key:  testKey,
+			id:   testID,
+			err:  nil,
+		},
+		{
+			desc: "Get thing ID from cache for non existing thing",
+			key:  "nonExistingKey",
+			id:   testID,
+			err:  errors.ErrNotFound,
+		},
+		{
+			desc: "Get thing ID from cache for empty key",
+			key:  "",
+			id:   testID2,
+			err:  nil,
```

This should return an error. We can check that key is always not empty before retrieving from redis